hooks/hookexecution/execution.go

```diff
@@ -0,0 +1,233 @@
+		}(hook, mCtx)
+	}
+
+	go func() {
+		wg.Wait()
+		close(resp)
+	}()
+
+	hookResponses := collectHookResponses(resp, done)
+
+	return processHookResponses(executionCtx, hookResponses, payload)
+}
+
+func executeHook[H any, P any](
+	moduleCtx hookstage.ModuleInvocationContext,
+	hw hooks.HookWrapper[H],
+	payload P,
+	hookHandler hookHandler[H, P],
+	timeout time.Duration,
+	resp chan<- hookResponse[P],
+	done <-chan struct{},
+) {
+	hookRespCh := make(chan hookResponse[P], 1)
+
+	select {
+	case <-done:
+		return
+	default:
+		startTime := time.Now()
+		hookId := HookID{hw.Module, hw.Code}
+
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+			result, err := hookHandler(ctx, moduleCtx, hw.Hook, payload)
+			hookRespCh <- hookResponse[P]{
+				Result: result,
+				Err:    err,
+			}
+		}()
+
+		select {
+		case res := <-hookRespCh:
+			res.HookID = hookId
+			res.ExecutionTime = time.Since(startTime)
+			resp <- res
+		case <-time.After(timeout):
+			resp <- hookResponse[P]{
+				Err:           TimeoutError{},
+				ExecutionTime: time.Since(startTime),
+				HookID:        hookId,
+				Result:        hookstage.HookResult[P]{},
+			}
+		case <-done:
+			return
+		}
+	}
+}
+
+func collectHookResponses[P any](
+	resp <-chan hookResponse[P],
+	done chan<- struct{},
+) []hookResponse[P] {
+	hookResponses := make([]hookResponse[P], 0)
+	for r := range resp {
+		hookResponses = append(hookResponses, r)
+		if r.Result.Reject {
+			close(done)
+			break
+		}
+	}
+
+	return hookResponses
+}
+
+func processHookResponses[P any](
+	executionCtx executionContext,
+	hookResponses []hookResponse[P],
+	payload P,
+) (GroupOutcome, P, groupModuleContext, *RejectError) {
+	stage := hooks.Stage(executionCtx.stage)
+	groupOutcome := GroupOutcome{}
+	groupOutcome.InvocationResults = make([]HookOutcome, 0, len(hookResponses))
+	groupModuleCtx := make(groupModuleContext, len(hookResponses))
+
+	for i, r := range hookResponses {
+		groupOutcome.InvocationResults = append(groupOutcome.InvocationResults, HookOutcome{
+			Status:        StatusSuccess,
+			HookID:        r.HookID,
+			Message:       r.Result.Message,
+			DebugMessages: r.Result.DebugMessages,
+			AnalyticsTags: r.Result.AnalyticsTags,
+			ExecutionTime: ExecutionTime{r.ExecutionTime},
+		})
+		groupModuleCtx[r.HookID.ModuleCode] = r.Result.ModuleContext
+
+		if r.ExecutionTime > groupOutcome.ExecutionTimeMillis {
+			groupOutcome.ExecutionTimeMillis = r.ExecutionTime
+		}
+
+		if r.Err != nil {
+			groupOutcome.InvocationResults[i].Errors = append(groupOutcome.InvocationResults[i].Errors, r.Err.Error())
+			switch r.Err.(type) {
+			case TimeoutError:
+				groupOutcome.InvocationResults[i].Status = StatusTimeout
+			case FailureError:
+				groupOutcome.InvocationResults[i].Status = StatusFailure
+			default:
+				groupOutcome.InvocationResults[i].Status = StatusExecutionFailure
+			}
+			continue
```

Comment:
The code flow is hard to follow with all of the continue statements. What does it mean to continue within this loop? Is there a way to rewrite this loop and/or use helper methods to make the code flow more obvious?

Reply:
Yeah I would agree with this comment

Reply:
We've done some refactoring of this function, please check if it's clearer now.